File: bundles/plugins/org.bonitasoft.studio.dependencies/src-test/resources/tomcat/webapps/bonita/portal/scripts/includes/table.js
/**
 * Copyright (C) 2011 BonitaSoft S.A. BonitaSoft, 32 rue Gustave Eiffel - 38000
 * Grenoble This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 2.0 of the License, or (at your option)
 * any later version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */

$(function() {

	$.tableKnownColumns = {
	    'checkbox' : [ 'checkboxes', 'checkbox' ],
	    'title' : [ 'title', 'name', 'firstname', 'firstName', 'first_name', 'lastName', 'last_name', 'displayname', 'displayName', 'display_name' ],
	    'description' : [ 'description', 'details' ],
	    'image' : [ 'image', 'thumbnail', 'icon', 'avatar' ],
	    'action' : [ 'action', 'actions' ]
	};

	// Plugin
	$.fn.prepareTable = function() {


		this.each(function() {
			var datatable = $(this);
			var table = $('.table', datatable);

			// get semantics
			var semantics = new Array();
			$('.thead .th', table).each(function() {
				var sem = $(this).attr('class').split(' ');
				for ( var i = 0; i < sem.length; i++) {
					if (sem[i].length > 3 && sem[i].substring(0, 3) == 'th_') {
						semantics.push(sem[i].substring(3));
						return;
					}
				}
				semantics.push(undefined);
			});

			// Line counters
			var cnt_lines = 1, cnt_cols = 0;
			$('.thead,.tbody,.tfoot', table).each(function() {
				cnt_lines = 1;
				$('.tr', $(this)).each(function() {
					var line = $(this);
					line.addClass(cnt_lines % 2 == 0 ? 'even' : 'odd').addClass('tr_' + (cnt_lines++));
					cnt_cols = 0;
					$('.td, .th', $(this)).each(function() {
						var e = $(this);
						e.addClass(cnt_cols % 2 == 0 ? 'even' : 'odd').addClass((e.is('.td') ? 'td' : 'th') + '_' + semantics[cnt_cols++]).addClass((e.is('.td') ? 'td' : 'th') + '_' + (cnt_cols));

						// Col types
						for ( var cat in $.tableKnownColumns) {
							for ( var i = 0; i < $.tableKnownColumns[cat].length; i++) {
								if (e.is('.td_' + $.tableKnownColumns[cat][i])) {
									e.addClass('td_type_' + cat);
									break;
								} else if (e.is('.th_' + $.tableKnownColumns[cat][i])) {
									e.addClass('th_type_' + cat);
									break;
								}
							}
						}

					});
					$('.td:last', line).addClass('td_last');
					$('.th:last', line).addClass('th_last');
				});
				$('.tr:last', table).addClass('tr_last');
			});

			// Columns types
			$('.td, .th').each(function() {
				var $currentCell = $(this);

			})
		});

		return this;
	};


	// Register maker
	$.uiManager.addMaker(function(c) {

		$('.datatable', c).prepareTable();

		if (c.is('.datatable')) {
			c.prepareTable();
		}
	}, 1)


	/**
	 * JQuery+ maker : hides 2 nodata
	 */
	$.uiManager.addMaker(function(context) {
		if (context.is(".subtasks") || context.is(".tasks") ) {
			if (context.is(".empty")) {
				context.show();
				if (context.parent().find(".subtasks:visible").length > 1) {
					context.hide();
				}
				if (context.parent().find(".tasks:visible").length > 1) {
					context.hide();
				}
			} else {
				context.parent().find(".subtasks.empty").hide();
				context.parent().find(".tasks.empty").hide();
			}
		}
	});


});
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
function SortableItemTable() {}

SortableItemTable.getSortedElement = function (ui){ return ui.item.get(0); }
SortableItemTable.getNextRow = function(elt){ return $(elt).next(".tr"); }

SortableItemTable.generateIndex = function(elt, childNodes, initialDomIndex){ 
	var APIid = DOMUtil.getAPIid(elt);
	var index = null;
	for(var i = 0; i< childNodes.length && index==null; i++){
		if(APIid ===  DOMUtil.getAPIid(childNodes[i])){
			if(initialDomIndex < i){
				index = SortableItemTable.shiftDown(i);
			}else{
				index = SortableItemTable.shiftUp(i);
			}
		}
	}
	return index;
}

SortableItemTable.getProfilePageIndex = function(row, initialDomIndex){ 
	var childNodes = $(row).siblings(".tr:not(.tr_1)").addBack();
	SortableItemTable.generateIndex(row, childNodes, initialDomIndex);
	return index;
}

SortableItemTable.getProfileFolderIndex = function(col, initialDomIndex){ 
	var childNodes = $(col).siblings(".tablePannel").addBack();
	return SortableItemTable.generateIndex(col, childNodes, initialDomIndex);
}

SortableItemTable.getInitialProfileFolderDomIndex = function(col){ 
	var colAPIId = DOMUtil.getAPIid(col);
	var childNodes = $(col).siblings(".tablePannel").addBack();
	return SortableItemTable.getProfileDomFromChildnodes(childNodes, colAPIId);
}
SortableItemTable.getProfileDomFromChildnodes = function(childNodes, searchedAPIID){
	for(var i = 0; i< childNodes.length; i++){
		if(searchedAPIID ===  DOMUtil.getAPIid(childNodes[i])){
			return i;
		}
	}
	return null;
}
SortableItemTable.getInitialProfilePageDomIndex = function(col){ 
	var colAPIId = DOMUtil.getAPIid(col);
	var childNodes = $(col).siblings(".tr:not(.tr_1)").addBack();
	return SortableItemTable.getProfileDomFromChildnodes(childNodes, colAPIId);
}

SortableItemTable.shiftUp = function(i){ 
	index = i == 0 ? -1 : ((i-1)*2)+1;
	return index;
}
SortableItemTable.shiftDown = function(i){ 
	index = i == 0 ? -1 : (i*2)+1;
	return index;
}

function DOMUtil() {}
DOMUtil.getAPIid= function(elt){
	var apiidPrefix = "APIID";
	classList = elt.className.split(/\s+/);
	var apiid = null;
	for (var i = 0; i < classList.length && apiid==null; i++) {
		if (classList[i].substring(0, apiidPrefix.length) === apiidPrefix) {
			apiid = classList[i].substring(apiidPrefix.length + 1);
		}
	}
	return apiid;
}

File: bundles/plugins/org.bonitasoft.studio.model/src/org/bonitasoft/studio/model/form/Table.java
/**
 * Copyright (C) 2009-2015 BonitaSoft S.A.
 * BonitaSoft, 32 rue Gustave Eiffel - 38000 Grenoble
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2.0 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.bonitasoft.studio.model.form;

import org.bonitasoft.studio.model.expression.Expression;

/**
 * <!-- begin-user-doc -->
 * A representation of the model object '<em><b>Table</b></em>'.
 * <!-- end-user-doc -->
 *
 * <p>
 * The following features are supported:
 * </p>
 * <ul>
 *   <li>{@link org.bonitasoft.studio.model.form.Table#isUsePagination <em>Use Pagination</em>}</li>
 *   <li>{@link org.bonitasoft.studio.model.form.Table#isAllowSelection <em>Allow Selection</em>}</li>
 *   <li>{@link org.bonitasoft.studio.model.form.Table#isSelectionModeIsMultiple <em>Selection Mode Is Multiple</em>}</li>
 *   <li>{@link org.bonitasoft.studio.model.form.Table#getMaxRowForPagination <em>Max Row For Pagination</em>}</li>
 *   <li>{@link org.bonitasoft.studio.model.form.Table#getColumnForInitialSelectionIndex <em>Column For Initial Selection Index</em>}</li>
 *   <li>{@link org.bonitasoft.studio.model.form.Table#getSelectedValues <em>Selected Values</em>}</li>
 * </ul>
 *
 * @see org.bonitasoft.studio.model.form.FormPackage#getTable()
 * @model
 * @generated
 */
public interface Table extends AbstractTable, MultipleValuatedFormField {
	/**
     * Returns the value of the '<em><b>Use Pagination</b></em>' attribute.
     * The default value is <code>"false"</code>.
     * <!-- begin-user-doc -->
	 * <p>
	 * If the meaning of the '<em>Use Pagination</em>' attribute isn't clear,
	 * there really should be more of a description here...
	 * </p>
	 * <!-- end-user-doc -->
     * @return the value of the '<em>Use Pagination</em>' attribute.
     * @see #setUsePagination(boolean)
     * @see org.bonitasoft.studio.model.form.FormPackage#getTable_UsePagination()
     * @model default="false"
     * @generated
     */
	boolean isUsePagination();

	/**
     * Sets the value of the '{@link org.bonitasoft.studio.model.form.Table#isUsePagination <em>Use Pagination</em>}' attribute.
     * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
     * @param value the new value of the '<em>Use Pagination</em>' attribute.
     * @see #isUsePagination()
     * @generated
     */
	void setUsePagination(boolean value);

	/**
     * Returns the value of the '<em><b>Allow Selection</b></em>' attribute.
     * The default value is <code>"false"</code>.
     * <!-- begin-user-doc -->
	 * <p>
	 * If the meaning of the '<em>Allow Selection</em>' attribute isn't clear,
	 * there really should be more of a description here...
	 * </p>
	 * <!-- end-user-doc -->
     * @return the value of the '<em>Allow Selection</em>' attribute.
     * @see #setAllowSelection(boolean)
     * @see org.bonitasoft.studio.model.form.FormPackage#getTable_AllowSelection()
     * @model default="false"
     * @generated
     */
	boolean isAllowSelection();

	/**
     * Sets the value of the '{@link org.bonitasoft.studio.model.form.Table#isAllowSelection <em>Allow Selection</em>}' attribute.
     * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
     * @param value the new value of the '<em>Allow Selection</em>' attribute.
     * @see #isAllowSelection()
     * @generated
     */
	void setAllowSelection(boolean value);

	/**
     * Returns the value of the '<em><b>Selection Mode Is Multiple</b></em>' attribute.
     * The default value is <code>"true"</code>.
     * <!-- begin-user-doc -->
	 * <p>
	 * If the meaning of the '<em>Selection Mode Is Multiple</em>' attribute isn't clear,
	 * there really should be more of a description here...
	 * </p>
	 * <!-- end-user-doc -->
     * @return the value of the '<em>Selection Mode Is Multiple</em>' attribute.
     * @see #setSelectionModeIsMultiple(boolean)
     * @see org.bonitasoft.studio.model.form.FormPackage#getTable_SelectionModeIsMultiple()
     * @model default="true"
     * @generated
     */
	boolean isSelectionModeIsMultiple();

	/**
     * Sets the value of the '{@link org.bonitasoft.studio.model.form.Table#isSelectionModeIsMultiple <em>Selection Mode Is Multiple</em>}' attribute.
     * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
     * @param value the new value of the '<em>Selection Mode Is Multiple</em>' attribute.
     * @see #isSelectionModeIsMultiple()
     * @generated
     */
	void setSelectionModeIsMultiple(boolean value);

	/**
     * Returns the value of the '<em><b>Max Row For Pagination</b></em>' containment reference.
     * <!-- begin-user-doc -->
	 * <p>
	 * If the meaning of the '<em>Max Row For Pagination</em>' containment reference isn't clear,
	 * there really should be more of a description here...
	 * </p>
	 * <!-- end-user-doc -->
     * @return the value of the '<em>Max Row For Pagination</em>' containment reference.
     * @see #setMaxRowForPagination(Expression)
     * @see org.bonitasoft.studio.model.form.FormPackage#getTable_MaxRowForPagination()
     * @model containment="true"
     * @generated
     */
	Expression getMaxRowForPagination();

	/**
     * Sets the value of the '{@link org.bonitasoft.studio.model.form.Table#getMaxRowForPagination <em>Max Row For Pagination</em>}' containment reference.
     * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
     * @param value the new value of the '<em>Max Row For Pagination</em>' containment reference.
     * @see #getMaxRowForPagination()
     * @generated
     */
	void setMaxRowForPagination(Expression value);

	/**
     * Returns the value of the '<em><b>Column For Initial Selection Index</b></em>' containment reference.
     * <!-- begin-user-doc -->
	 * <p>
	 * If the meaning of the '<em>Column For Initial Selection Index</em>' containment reference isn't clear,
	 * there really should be more of a description here...
	 * </p>
	 * <!-- end-user-doc -->
     * @return the value of the '<em>Column For Initial Selection Index</em>' containment reference.
     * @see #setColumnForInitialSelectionIndex(Expression)
     * @see org.bonitasoft.studio.model.form.FormPackage#getTable_ColumnForInitialSelectionIndex()
     * @model containment="true"
     * @generated
     */
	Expression getColumnForInitialSelectionIndex();

	/**
     * Sets the value of the '{@link org.bonitasoft.studio.model.form.Table#getColumnForInitialSelectionIndex <em>Column For Initial Selection Index</em>}' containment reference.
     * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
     * @param value the new value of the '<em>Column For Initial Selection Index</em>' containment reference.
     * @see #getColumnForInitialSelectionIndex()
     * @generated
     */
	void setColumnForInitialSelectionIndex(Expression value);

	/**
     * Returns the value of the '<em><b>Selected Values</b></em>' containment reference.
     * <!-- begin-user-doc -->
	 * <p>
	 * If the meaning of the '<em>Selected Values</em>' containment reference isn't clear,
	 * there really should be more of a description here...
	 * </p>
	 * <!-- end-user-doc -->
     * @return the value of the '<em>Selected Values</em>' containment reference.
     * @see #setSelectedValues(Expression)
     * @see org.bonitasoft.studio.model.form.FormPackage#getTable_SelectedValues()
     * @model containment="true"
     * @generated
     */
	Expression getSelectedValues();

	/**
     * Sets the value of the '{@link org.bonitasoft.studio.model.form.Table#getSelectedValues <em>Selected Values</em>}' containment reference.
     * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
     * @param value the new value of the '<em>Selected Values</em>' containment reference.
     * @see #getSelectedValues()
     * @generated
     */
	void setSelectedValues(Expression value);

} // Table
